Sources/SourceKitLSP/SourceKitLSPServer.swift

```diff
@@ -209,29 +209,40 @@ package actor SourceKitLSPServer {
     guard var url = uri.fileURL?.deletingLastPathComponent() else {
       return nil
     }
-    let projectRoots = await self.workspacesAndIsImplicit.filter { !$0.isImplicit }.asyncCompactMap {
+
+    // Roots of opened workspaces - only consider explicit here (all implicit must necessarily be subdirectories of
+    // the explicit workspace roots)
+    let workspaceRoots = workspacesAndIsImplicit.filter { !$0.isImplicit }.compactMap { $0.workspace.rootUri?.fileURL }
+
+    // We want to skip creating another workspace if the project roots are the same. This could happen if an existing
+    // workspace hasn't reloaded after a new file was added to it (and thus that build system needs to be reloaded).
+    let projectRoots = await workspacesAndIsImplicit.asyncCompactMap {
       await $0.workspace.buildSystemManager.projectRoot
     }
-    let rootURLs = workspacesAndIsImplicit.filter { !$0.isImplicit }.compactMap { $0.workspace.rootUri?.fileURL }
-    while url.pathComponents.count > 1 && rootURLs.contains(where: { $0.isPrefix(of: url) }) {
+
+    while url.pathComponents.count > 1 && workspaceRoots.contains(where: { $0.isPrefix(of: url) }) {
       defer {
         url.deleteLastPathComponent()
       }
-      // Ignore workspaces that have the same project root as an existing workspace.
-      // This might happen if there is an existing SwiftPM workspace that hasn't been reloaded after a new file
-      // was added to it and thus currently doesn't know that it can handle that file. In that case, we shouldn't open
-      // a new workspace for the same root. Instead, the existing workspace's build system needs to be reloaded.
+
       let uri = DocumentURI(url)
-      guard let buildSystemSpec = determineBuildSystem(forWorkspaceFolder: uri, options: self.options) else {
+      let options = SourceKitLSPOptions.merging(base: self.options, workspaceFolder: uri)
+
+      // Some build systems consider paths outside of the folder (eg. BSP has settings in the home directory). If we
+      // allowed those paths, then the very first folder that the file is in would always be its own build system - so
+      // skip them in that case.
```

Comment:
Comment out of date if `onlyConsiderRoot` is not used?

Reply:
Whoops, it's... not out of date. I just forgot to actually pass it through into searching for BSP. No test for it because it's all environment variables but given this, probably need a way to test that 🤔

Reply:
Noticed some more fun when looking at this. Today we can have:
1. Explicit workspace + compilation db build system using build/compile_commands.json
2. File unhandled in build directory
3. Implicit workspace created as the roots don't match up (the explicit project root doesn't include build, but the implicit will)

IMO it's also a little confusing that we just ignore an invalid DB. Might change project root to just check for existence + return the directory that the DB is in rather than the directory it was searched from.